Source Mapping

alanmcruickshank opened the pull request:
This needs a few more test cases, but is otherwise ready for review. I'll add test cases before merging.

This is also another **BIT OF A MONSTER PR**. Sorry that this will be quite hard to review, but it's quite a deep reaching change in how templating and fixing is done.

This fixes #273 and relates to #368.

In short, this generates a source map from the template, this allows us to keep a reference for where in the template each element of the parsed file comes from. We can then use that mapping in two ways:
- We can use it when displaying linting errors so that they are displayed using the position in the *source* file rather than in the version _post-templating_. See #273 
- We can use it when fixing to allow us to apply fixes more intelligently, and eliminate the need for `difflib` or `cdifflib` which should increase the speed of fixing operations significantly. See #368 

This is still fresh off the press and so highly likely that there are optimisations to be had in the code, it's also possible that I've been staring at this for far too long and have done something which could be much simpler. Feedback would be great on:
1. The overall approach.
2. Logic location. I've tried to be thoughtful about where I put the logic for different things, but would love input on how well I've achieved something sensible.
3. Low level style. This is just a big PR, any low level feedback gratefully received.
4. Suggestions on tests. As above, I think this needs more tests - in particular I already know I need some tests for templated queries which try to apply fixes *within a loop* (which I have a feeling will currently break this code). Any other suggestions for more tests would be excellent.

alanmcruickshank commented:
Nice. Looking at the benchmark tests, the time to complete `B_002_pearson` has basically halved. @barrywhart , that test is based on your large query, so I'm hoping that this should mean significant performance increases for your use case. There's probably some headroom for more performance gains within this code too.

pwildenhain commented:
Does the work in this PR also open up the possibility of adding an option to ignore templated lines when linting?

alanmcruickshank commented:
> Does the work in this PR also open up the possibility of adding an option to ignore templated lines when linting?

I think so yes. Perhaps more excitingly I think it *might* open up the possibility of explicitly enforcing indentation around templated sections. I'll make a point of exploring that in the test cases. I think I'll have to do most of the work on that to stop some weird behavior around jinja tags anyway.

If I can put together an example of linting the indentation of templated sections too then I'll add it as a configurable feature.

pwildenhain commented:
For linting the indentation of jinja blocks, are you thinking about implementing a separate rule? Or maybe refactoring `Rule_L003` to be extendable, much like `Rule_L011`?


Super pumped about this BTW 😁

NiallRees commented on src/sqlfluff/core/linter.py:
```diff
@@ -105,275 +100,161 @@ def is_clean(self):
+                insert_buff = ""
+                for seg_idx, segment in enumerate(seg.segments):
+                    # First check for insertions.
+                    # We know it is if the position marker is NOT ENRICHED.
```
We know it is what?

NiallRees commented on src/sqlfluff/core/parser/markers.py:
```diff
@@ -3,15 +3,53 @@
+    def from_fresh(cls):
+        """Construct a fresh position marker.
+
+        This should be the default way of creating a new postion marker, and
```
```suggestion
        This should be the default way of creating a new position marker, and
```

NiallRees commented on src/sqlfluff/core/templaters/python.py:
```diff
@@ -0,0 +1,526 @@
+
+    @staticmethod
+    def infer_type(s):
+        """Infer a python type from a string ans convert.
```
```suggestion
        """Infer a python type from a string and convert.

```

NiallRees commented on src/sqlfluff/core/templaters/python.py:
```diff
@@ -0,0 +1,526 @@
+        raw_sliced = list(cls._slice_template(raw_str))
+        # Find the literals
+        literals = [elem[0] for elem in raw_sliced if elem[1] == "literal"]
+        # Calculate occurances
```
```suggestion
        # Calculate occurrences

```

NiallRees commented on src/sqlfluff/core/templaters/python.py:
```diff
@@ -0,0 +1,526 @@
+        # Find the literals
+        literals = [elem[0] for elem in raw_sliced if elem[1] == "literal"]
+        # Calculate occurances
+        raw_occurances = cls._substring_occurances(raw_str, literals)
```
```suggestion
        raw_occurrences = cls._substring_occurrences(raw_str, literals)

```
etc

NiallRees commented:
Have scanned through all but tests and looks good. I still have no idea how you just churn this stuff out 😂. Will check the tests in a bit, and like @pwildenhain says I'm very excited about the speed improvements and to get rid off difflib. The logic in linter.py was hard to debug, it now looks far more approachable. Also super excited about what this means for https://github.com/sqlfluff/sqlfluff/issues/464 and other templating related rules. Great work!

alanmcruickshank commented:
> For linting the indentation of jinja blocks, are you thinking about implementing a separate rule? Maybe refactoring `Rule_L003` to be extendable, much like `Rule_L011`?

I think I'll extend L003 a little. Almost all of the indentation logic is in there, so seems silly to move it. Planning to use the configurable meta segments to make a kind of _jinja tag placeholder_ tag which represents the tag itself but has no length in the templated file. Should be able to combine that with configured Indent and Dedent tags generated at templating time to add expected indents into the file. From there the existing logic *should* "just work". 🤞 [famous last words].

> Have scanned through all but tests and looks good. I still have no idea how you just churn this stuff out 😂.

I had a brainwave on this one last week - so spent a chunk of the weekend on it. Lots of thinking time in advance is the key to swift execution. It's also much easier for me to do these kind of PRs on this codebase because I wrote most of the rest so I already know where the seams are. 😁

pwildenhain commented on test/core/templaters/python_test.py:
```diff
@@ -0,0 +1,25 @@
+from sqlfluff.core import SQLTemplaterError
+
+
+PYTHON_STRING = "SELECT * FROM {blah}"
```
Is there a reason you're using global variables rather than fixtures for the test strings?

alanmcruickshank replied:
I'm using them because I'm using them in parameterized tests. Do you know if you can use fixtures in parameterised tests? If you do, then I'm totally up for it.

The idea would be to use one fixture for some tests and another fixture for others (or maybe some tests don't need the fixture).

Getting a good approach on this would be excellent to reduce bloat on some of the tests in `templaters.base_test,py` which is now _pretty ugly_.

barrywhart replied:
Are we talking about PyTest? If so, I'm pretty sure you can use fixtures in parameterized tests.

pwildenhain replied:
Hmm, I just spend some time researching, and nothing turned up. Given that, I'm onboard with this approach

pwildenhain commented on src/sqlfluff/core/templaters/python.py:
```diff
@@ -80,3 +82,85 @@ def process(self, in_str, fname=None, config=None):
+
+    @staticmethod
+    def _findall(substr: str, in_str: str) -> Iterator[int]:
+        """Yields all the positions sbstr within in_str.
```
`sbstr ` --> `substr`

pwildenhain commented on src/sqlfluff/core/templaters/python.py:
```diff
@@ -80,3 +82,85 @@ def process(self, in_str, fname=None, config=None):
+                        # We double them here to make the raw
+                        (literal_text[first_char[1] : idx] * 2, "escaped", in_idx)
+                    )
+                    # Will always be 2 in this case
+                    in_idx += 2
```
Why will it always be 2?

alanmcruickshank replied:
Because all python escape sequences are two characters long. I've added a comment to that effect.

pwildenhain commented on src/sqlfluff/core/templaters/__init__.py:
```diff
@@ -6,6 +6,6 @@
-from .base import RawTemplateInterface
-from .jinja import JinjaTemplateInterface
-from .python import PythonTemplateInterface
+from .base import RawTemplater
+from .jinja import JinjaTemplater
+from .python import PythonTemplater
```
👍 for the more concise names

pwildenhain reviewed (commented):
Looking good so far! I didn't make it all the way through, but this is a very impressive refactor, and looks like it makes handling templates a much nicer experience.

Any thoughts on how the work in here might correlate with what's going on in #508? Or would it not really matter, since `dbt` is really just a `jinja` templated file when it comes down to it.

alanmcruickshank commented on src/sqlfluff/core/templaters/python.py:
```diff
@@ -0,0 +1,81 @@
+
+
+@register_templater
+class PythonTemplateInterface(RawTemplateInterface):
```
I don't think anyone uses it. It's a useful step on the way to the python jinja templater. Someone might want to use it in future 🤷

alanmcruickshank commented:
@barrywhart - I've just tested the slow query from #368 on this branch [which also includes the changes from #526], and running fix takes around 22s (including time for key presses). Can you confirm you're getting the same results?

NiallRees commented:
> @barrywhart - I've just tested the slow query from #368 on this branch [which also includes the changes from #526], and running fix takes around 22s (including time for key presses). Can you confirm you're getting the same results?

I just tested the query with the command given in #368 and I'm getting 14 seconds incl key press and fixes 🎉

NiallRees commented:
I just compared this with the master branch on pet.sql and whole process took 1:02, compared with 1:39 before. The time to lint took basically the same time, but applying the fixes was pretty instantaneous (sub-second).

dmateusp commented on src/sqlfluff/core/linter.py:
```diff
@@ -105,275 +101,136 @@ def is_clean(self):
+
+        # Make sure no patches overlap and divide up the source file into slices.
+        # Any Template tags in the source file should be "untouchable".
+        untouchable_slices = self.templated_file.untouchable_slices()
```
Not sure about the `untouchable_slice` concept, if it is only for templated parts, couldn't we name it `template_slice`?

alanmcruickshank replied:
Ha - yeah that's probably a much better name! I'll see what I can do on that front.

alanmcruickshank replied:
On reflection, it's not just _templated slices_, but a subset of them which should be expected to have zero-length in the templated file (i.e. comments or blocks). I've renamed the method to `source_only_slices` and updated the signature to return `RawFileSlice` objects directly.

alanmcruickshank commented:
@NiallRees @barrywhart , the benchmark test which is modeled after the long fix query from #368 is taking ~21s on the most recent commit, and back around the last release the same benchmark was taking ~53s, so that feels like a pretty good performance boost. Given @NiallRees 's comment on fixing being basically instant now, I think this source mapping PR is doing it's job and we've successfully moved the bottleneck to a different place. 🏆 

I've done most of the tidy up I want to do with this now. I'll do a few docs updates, but then this will be ready for a full review.

dmateusp commented on src/sqlfluff/core/linter.py:
```diff
@@ -105,275 +101,136 @@ def is_clean(self):
 
-            linter_logger.debug(
-                "%04d: Blocks: template:%s, fix:%s", loop_idx, templ_block, fixed_block
+        # We now convert enrich the patches into source space
```
Can you clarify that comment? Does that mean we're applying the patches?

alanmcruickshank replied:
Good feedback. I've edited this comment to `We now convert templated-space patches into source-space patches.`.

What we're doing is taking a load of patches which are referenced using positions in the templated file to ones which are referred to in the source file, which means we can dedupe, and also apply them around templated sections.

dmateusp commented on src/sqlfluff/core/linter.py:
```diff
@@ -105,275 +101,136 @@ def is_clean(self):
-        # Compare Fixed with Templated
-        diff_fix = SequenceMatcher(
-            autojunk=None, a=self.file_mask[1], b=self.file_mask[2]
+        # The sliced file is contigious in the TEMPLATED space.
```
I'm a bit confused by the difference between a "Slice" and a "Segment"

alanmcruickshank replied:
I'll clarify that in the docstring above, but the thrust is that a `Slice` is something which is worked out _before_ lexing, whereas a `Segment` is something that is post-lexing. Practically, Segments will probably be subdivisions of Slices.

alanmcruickshank replied:
From the new docstring:

```
There is an important distinction here between Slices and
Segments. A Slice is a portion of a file which is determined
by the templater based on which portions of the source file
are templated or not, and therefore before Lexing and so is
completely dialect agnostic. A Segment is determined by the
Lexer from portions of strings after templating.
```

Is that clear enough, or is it worth me going into more detail?

dmateusp replied:
I think it's clear!

dmateusp commented on src/sqlfluff/core/parser/lexer.py:
```diff
@@ -295,7 +300,7 @@ def __init__(
         )
 
     def lex(self, raw: str) -> Tuple[Tuple[RawSegment, ...], List[SQLLexError]]:
```
```suggestion
    def lex(self, raw: Union[str, TemplatedFile]) -> Tuple[Tuple[RawSegment, ...], List[SQLLexError]]:
```
from your docstring below

dmateusp commented on src/sqlfluff/core/parser/lexer.py:
```diff
@@ -322,9 +330,111 @@ def lex(self, raw: str) -> Tuple[Tuple[RawSegment, ...], List[SQLLexError]]:
+        in the source file.
+        """
+        # Check this is actually a templated file.
+        if not isinstance(templated_file, TemplatedFile):
```
should that check go outside of the function?

The arg and the docs specify that this function takes a `TemplatedFile` but it has the "hidden" behavior of returning the input unchanged if it's not

alanmcruickshank replied:
Good idea 👍

dmateusp commented on src/sqlfluff/core/parser/lexer.py:
```diff
@@ -322,9 +330,111 @@ def lex(self, raw: str) -> Tuple[Tuple[RawSegment, ...], List[SQLLexError]]:
+                    break
+                elif ut_slice.start == source_slice.start:
+                    lexer_logger.debug(
+                        "Found untouchable slot! %s, %s, %s",
```
What is a slot? could it be a section?

dmateusp commented on src/sqlfluff/core/parser/markers.py:
```diff
@@ -3,16 +3,64 @@
+        This should be the default way of creating a new position marker, and
+        is where we define what the `start` of a file looks like.
+        """
+        return cls(1, 1, 1, 0)
```
Should these be defaults in `__init__` instead? and then the method can be removed

alanmcruickshank replied:
🤦 - Good call. This is some *VERY* old code, which I've not really revisited in a while. 👍

dmateusp commented on src/sqlfluff/core/parser/segments/base.py:
```diff
@@ -767,15 +836,18 @@ def apply_fixes(self, fixes):
-                    if seg.is_meta:
-                        seg_buffer.append(seg)
-                        continue
+                    # if seg.is_meta:
```
fyi commented code here

dmateusp commented on src/sqlfluff/core/parser/segments/base.py:
```diff
@@ -842,42 +914,100 @@ def realign(self):
+            linter_logger.debug("Yeilding literal patch: %r", patch)
+            yield patch
+        else:
+            # This segment isn't a literal, but has changed, we need to go deeper.
```
I'm with you <img src="https://i.kym-cdn.com/entries/icons/mobile/000/012/886/wntgd.jpg" height=150 />

alanmcruickshank replied:
I thought someone might enjoy that reference 😆

dmateusp commented on src/sqlfluff/core/parser/segments/indent.py:
```diff
@@ -1,30 +1,24 @@
-    be compared later.
-    """
+class MetaSegment(RawSegment):
+    """A segment which is empty but indicates where something should be."""
```
Add a note about how it's used as a placeholder for pre-templated code?

alanmcruickshank replied:
Excellent idea, I think this is also related to your comment before of:

> There are some pre-templating rules like "have a space between {{ and }} and the expression" which I'm not sure could be supported with the slices?

I'll extend the same so that it contains a reference to the source string in case anyone wants to lint it.

dmateusp commented on src/sqlfluff/core/rules/base.py:
```diff
@@ -107,7 +107,20 @@ def __init__(self, edit_type, anchor, edit=None):
+            edit = [edit]
         self.edit = edit
+        if self.edit:
+            # Copy all the elements of edit to stop contamination
```
🚫 🦠

alanmcruickshank replied:
I've added some more detailed comments here to explain why...

```
        # Copy all the elements of edit to stop contamination.
        # We're about to start stripping the position markers
        # of some of the elements and we don't want to end up
        # stripping the positions of the original elements of
        # the parsed structure.
        self.edit = copy.deepcopy(edit)
```

dmateusp commented on src/sqlfluff/core/rules/base.py:
```diff
@@ -107,7 +107,20 @@ def __init__(self, edit_type, anchor, edit=None):
         self.edit = edit
+        if self.edit:
+            # Copy all the elements of edit to stop contamination
+            self.edit = [copy.copy(elem) for elem in edit]
```
Can't you `deepcopy` the list directly?

alanmcruickshank replied:
Good idea 👍

dmateusp commented on src/sqlfluff/core/templaters/base.py:
```diff
@@ -0,0 +1,359 @@
+templater_logger = logging.getLogger("sqlfluff.templater")
+
+
+def templater_selector(s=None, **kwargs):
```
This should probably by `s="jinja"` and we can remove `s = s or "jinja"`, that way it's more information from the function signature

alanmcruickshank replied:
It's set up that way so that if `None` is explicitly passed at `s` then we still get the jinja templater. The way config is set up, we always get the keyword, but sometimes with `None`. I didn't want to have the default for `jinja` set it several different places, and so this approach means we can set it in just one place and still get the desired effect.

dmateusp commented on src/sqlfluff/core/templaters/base.py:
```diff
@@ -0,0 +1,359 @@
+        )
+
+
+def register_templater(cls):
```
Is this more related to making templaters "plugins" rather than "source mapping"?

alanmcruickshank replied:
This is old code, that is just being moved around in this PR. I think this may get refactored as part of the plugin work, but that's a job for another day.

dmateusp commented on src/sqlfluff/core/templaters/jinja.py:
```diff
@@ -1,168 +1,21 @@
 
 @register_templater
-class JinjaTemplateInterface(PythonTemplateInterface):
+class JinjaTemplater(PythonTemplater):
```
👍 for the naming change :)

dmateusp reviewed (commented):
This is indeed a huge PR haha! But it's very exciting 🎉 

Conceptually I'd like to understand the difference between a Slice and a Segment, any reason we can't just nest the templated segments inside a "TemplateSegment" ?

There are some pre-templating rules like "have a space between `{{` and `}}` and the expression" which I'm not sure could be supported with the slices? I think we need to branch the AST with something like:
```python
class TemplateSegment:
  raw: str
  templated_segment: Segment 
```

alanmcruickshank commented on src/sqlfluff/api/simple.py:
```diff
@@ -73,7 +73,7 @@ def parse(sql, dialect="ansi"):
     sql = _unify_str_or_file(sql)
     linter = Linter(dialect=dialect)
-    parsed, violations, _ = linter.parse_string(sql)
+    parsed, violations, _, _ = linter.parse_string(sql)
```
Great idea. I'll ping you on the code to get your review on what I've done.

alanmcruickshank commented:
@dmateusp - Thanks for the detailed review, could you re-review the changes since, especially my treatment of:
- the new `ParsedString` class to hold the result of `.parse_string()`
- the rename of the placeholder to `TemplateSegment`, including how both the type and source str and held, hopefully in a way that facilitates future linting.

@sqlfluff/sqlfluff-maintainers - I've added a couple of more complex jinja examples, but I'd be really grateful if anyone could test this branch against some of the more hairy jinja code you might have to hand. Ideally suggesting a couple of test cases would be excellent to add to the two major ones I've already added: `012_L006_templating` & `011_L003L006_indentation_3`.

barrywhart commented:
We've done some fairly complex Jinja templating involving loops and conditionals. I'll look for some examples to share tomorrow. Does the SQLFluff config file support complex types such as arrays? I think we added that awhile back. I think that may make it easier to use the examples in the tests.

barrywhart commented:
@alanmcruickshank: I ran this branch on the query from #368, and it takes 11 seconds. Very nice!

barrywhart commented:
I didn't have time to test the branch with this, but I'm providing this complex, templated SQL query in case it seems like a useful test case.

Query template:
```
WITH
  raw_effect_sizes AS (
  SELECT
    COUNT(1) AS campaign_count,
    {{corr_states}}
    {% for action in considered_actions %}
    ,SAFE_DIVIDE(SAFE_MULTIPLY(CORR({{metric}}_rate_su, {{action}}), STDDEV_POP({{metric}}_rate_su)), STDDEV_POP({{action}})) AS {{metric}}_{{action}}
    {% endfor %}
  FROM
    `{{gcp_project}}.{{dataset}}.global_actions_states`
  GROUP BY
    {{corr_states}}
  ),
  
  {% for action in considered_actions %}
  {{action}}_raw_effect_sizes AS (
  SELECT
    COUNT(1) AS campaign_count_{{action}},
    {{corr_states}}
    ,SAFE_DIVIDE(SAFE_MULTIPLY(CORR({{metric}}_rate_su, {{action}}), STDDEV_POP({{metric}}_rate_su)),
      STDDEV_POP({{action}})) AS {{metric}}_{{action}}
  FROM
    `{{gcp_project}}.{{dataset}}.global_actions_states`
  WHERE
    {{action}} != -1
  GROUP BY
    {{corr_states}}
  ),
  {% endfor %}
  
  new_raw_effect_sizes AS (
  SELECT
    {{corr_states}}
    {% for action in considered_actions %}
    ,{{metric}}_{{action}}
    ,campaign_count_{{action}}
    {% endfor %}
  FROM
    {% for action in considered_actions %}
    {% if loop.first %}
    {{action}}_raw_effect_sizes
    {% else %}
  JOIN
    {{action}}_raw_effect_sizes
  USING
    ({{corr_states}})
    {% endif %}
    {% endfor %}
  ),

  imputed_effect_sizes AS (
  SELECT
    {{corr_states}}
    ,o.campaign_count AS campaign_count
    {% for action in considered_actions %}
    ,COALESCE(IF(IS_NAN(o.{{metric}}_{{action}}), 0, o.{{metric}}_{{action}}), 0) AS {{metric}}_{{action}}
    ,COALESCE(IF(IS_NAN(n.{{metric}}_{{action}}), 0, n.{{metric}}_{{action}}), 0) AS new_{{metric}}_{{action}}
    ,n.campaign_count_{{action}}
    {% endfor %}
  FROM
    raw_effect_sizes o
  JOIN
    new_raw_effect_sizes n
  USING
    ({{corr_states}})
  ),

  action_states AS (
  SELECT 
    {{action_states}}
  FROM `{{gcp_project}}.{{dataset}}.global_state_space`
  GROUP BY {{action_states}})

SELECT
  imputed_effect_sizes.*,
  {{action_states}}
FROM
  imputed_effect_sizes
CROSS JOIN action_states
```

Template parameters:
```
{'action_states': 'has_used_small_subject_line\n'
                  ' ,has_used_personal_subject\n'
                  ' ,has_used_personal_to\n',
 'considered_actions': ['uses_small_subject_line',
                        'uses_personal_subject',
                        'uses_personal_to'],
 'corr_states': 'state_user_v_peer_open\n ,business_type',
 'dataset': 'my_dataset',
 'gcp_project': 'my_project',
 'metric': 'open'}
```

alanmcruickshank commented:
@barrywhart - I just ran that query through the current templating a fixing logic. That's a real monster. Plenty to chew on here. Excellent test case! 🤣 🤣 🤣 🤣 🤣 🤣 🤣 

Will report back in due course with a damage report.

barrywhart commented:
Heh, I'm happy to oblige. We might have some worse ones, but I'm proud of this find!

dmateusp commented on CHANGELOG.md:
```diff
@@ -22,6 +22,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  For loop or block tags, they also generate an `Indent` and
+  `Dedent` tag accordingly (which can be enabled or disabled)
+  with a configuration value so that indentation around these
+  functions can be linted accordingly.
```
Are we running the linter twice? Once with the placeholders and once with the fully expanded SQL ?

alanmcruickshank replied:
No. We only run it on the fully expanded SQL. The placeholders will be within the fully expanded sql, but they have zero length (i.e. their `.raw()` element is a zero length string `""`). This is very similar to how I've implemented indentation, where `Indent` and `Dedent` segments which also have zero length.

If you run `sqlfluff parse` on a basic file with some simple templating in it you'll see the parsed structure including any placeholders. That is the structure which is linted.

alanmcruickshank replied:
Do you think the current docs explain that well? I wrote the docs in parallel with the code, and I'm often revising large patches of this code at the moment so it's very possible that the docs were well intentioned at the time but now need to be a bit clearer.

dmateusp replied:
This Changelog note is clear, however I'm thinking for a user coming from DBT trying to implement some rules that apply pre-templating, like the issues that @NiallRees has been opening (and I'm also interested in).

Does this work belong in a different PR? How do you think that would work? Would we end up with 2 ASTs? Would we have to tag issues to be "pre-templating" "post-templating" or "both" so they are selectively applied? <- Those are the ones keeping me up at night :)

alanmcruickshank replied:
I've added some stubs into the code to open up the capability to fix certain things like the rules you've mentioned. There's a `TODO` note in the `linter.py`. But yes, this PR is already a monster. That is a job for another day.

If @NiallRees is watching 😄 : My intent is that a `TemplatedSegment` could have a `fixed_str` attribute, that when found would be used to replace the original version of the marker. It won't work in all cases - the way I've lexed templated code means that on very complicated jinja examples, I have to start making approximations, and in some of those we never get a `TemplatedSegment` in the lexed structure (so far it looks like I'm particularly bad at recognizing if statements).

dmateusp commented on CHANGELOG.md:
```diff
@@ -30,6 +38,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   needing to worry about wierd output or without the log handing getting
   in the way of your project.
+- Linting errors in the final file are now reported with their position
+  in the source file rather than in the templated file. This means
```
Would it be useful to report both the source file and target file lines? I imagine for big macros you would want the possibility to look at the compiled file (I do this today)

alanmcruickshank replied:
Potentially yes, I totally see your point. I'm not quite sure _how_ though. From a data structure point of view, the source and templated positions are both accessible within the new enriched position markers. We're just choosing to display one over the other. I briefly considered making it something configurable - would you make use of that feature if I made it a flag in the config file?

If running in a more verbose mode, you'll also already get an output of the compiled file in the logs if you want it for debugging, but I'm concious that you *also* get a lot of other logging. In big files it's a very very long log.

Totally up for suggestions on how to make this more useful!

dmateusp replied:
Do you think it would be feasible to compare the Line number and position of the source vs templated file and if they differ add a column in the output?

Like:
```
L:   3 | P:   5 | TL: 7 | TP: 10 | L027 | Unqualified reference 'current_timestampp' found in
                        | select with more than one referenced table/view.
```

Where TL and TP stand for Templated Line and Templated Position

barrywhart replied:
I like the TL / TP idea! Or something along those lines.

This is kind of "off topic", but at my job, we use the `pygments` package to output SQL errors (parse errors, etc.) with color syntax highlighting. I wonder if something similar could be a good feature for SQLFluff at some point. It has helped a bit with being able to quickly read and interpret error output.

alanmcruickshank replied:
I like both of these ideas, but I think both need fleshing out and we should make seperate issues for them.

@barrywhart - could you make one for the colour coded errors idea. In particular what you think should determine the colours?

@dmateusp - could you make another one for extending the position output to something with more detail in it? In my mind I'm not sure how someone would *use* the templated position, because that's not what they're actually looking at in an editor - but I think there's definitely something in there, because if users look closely at the moment there are *some* elements where I can't conclusively identify their source position and so display the templated position only, and that will likely confuse users at some point.

dmateusp replied:
Not sure I understand the first part of your question ^^', but after thinking more about this, I think it deserves some experimentation and focused discussion, worthy of its own PR

barrywhart replied:
Sure, I'll create a ticket. As a first approximation, it'd be "do what Pygments does now, but using the SQLFluff parser" (since it has more support for the carrots dialects of SQL).

dmateusp commented on src/sqlfluff/api/simple.py:
```diff
@@ -68,13 +68,12 @@ def parse(sql, dialect="ansi"):
-    if violations:
-        raise violations[0]
+    if parsed.violations:
+        raise parsed.violations[0]
```
Interesting! I didn't know we were only raising the first violation for parsing, I wonder if returning an Exception which encompasses all of the violations would be useful 🤔  (out of scope of the PR)

alanmcruickshank replied:
For parsing through the command line we display all of them. It's only in the simple api which we don't. I wanted something which provided a straightforward API which was easy to use, but that for more advanced use cases, developers would start to use the main `Linter()` class directly.

I think your idea of raising an exception with all of them is a strong one! Can you make a github issue for that with what you had in mind?

dmateusp replied:
created https://github.com/sqlfluff/sqlfluff/issues/569

dmateusp commented on src/sqlfluff/core/linter.py:
```diff
@@ -2,38 +2,44 @@
-        ["path", "violations", "time_dict", "tree", "file_mask", "ignore_mask"],
-    )
-):
+class ParsedString(NamedTuple):
```
🎉

dmateusp commented on src/sqlfluff/core/linter.py:
```diff
@@ -105,275 +111,145 @@ def is_clean(self):
+        NB: This is MUCH FASTER than the original approach
+        using difflib in pre 0.4.0.
 
+        There is an important distinction here between Slices and
```
:+1:

dmateusp commented on src/sqlfluff/core/parser/segments/indent.py:
```diff
@@ -105,4 +118,33 @@ class Dedent(Indent):
+
+
+class TemplateSegment(MetaSegment):
+    """A segment which is empty but indicates something should be.
```
Just wanted to be clear on the docstring here. How come the `source_str` is optional? Is there ever cases where we don't store the original string?

alanmcruickshank replied:
Goooood question. I'm not sure there is. Maybe I should revisit that. 🤔

alanmcruickshank replied:
This now raises an exception if not set. I think we'll rely on it later for some of the rules we want to do so best to make sure we can rely on it.

dmateusp reviewed (commented):
I had a quick second look! I just have some conceptual questions raised by what I read in the documentation

pwildenhain commented:
I gave this branch a go on a big templated block I'm working with, and it looks like it parsed it just fine 👌 

Here's my test file if you want to see how it parses it
<details>

```sql
    {% set
    neo_metrics = [
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Effective and Efficient Care'",
            "metric_name": "'Median NICU LOS'",
            "drill_down_1": "cohort_group",
            "drill_down_2": "admission_source",
            "metric_month": "episode_end_month",
            "num": "nicu_los_days",
            "denom": "null",
            "num_calculation": "'median'",
            "denom_calculation": "null",
            "metric_type": "'median'",
            "direction": "'down'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_episodes')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Effective and Efficient Care'",
            "metric_name": "'25th percentile NICU LOS'",
            "drill_down_1": "cohort_group",
            "drill_down_2": "admission_source",
            "metric_month": "episode_end_month",
            "num": "nicu_los_days",
            "denom": "null",
            "num_calculation": "'percentile_25'",
            "denom_calculation": "null",
            "metric_type": "'percentile'",
            "direction": "'down'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_episodes')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Effective and Efficient Care'",
            "metric_name": "'75th percentile NICU LOS'",
            "drill_down_1": "cohort_group",
            "drill_down_2": "admission_source",
            "metric_month": "episode_end_month",
            "num": "nicu_los_days",
            "denom": "null",
            "num_calculation": "'percentile_75'",
            "denom_calculation": "null",
            "metric_type": "'percentile'",
            "direction": "'down'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_episodes')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Effective and Efficient Care'",
            "metric_name": "'Median Hospital LOS'",
            "drill_down_1": "cohort_group",
            "drill_down_2": "admission_source",
            "metric_month": "hospital_discharge_month",
            "num": "hospital_los_days",
            "denom": "null",
            "num_calculation": "'median'",
            "denom_calculation": "null",
            "metric_type": "'median'",
            "direction": "'down'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_visits')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Effective and Efficient Care'",
            "metric_name": "'25th percentile Hospital LOS'",
            "drill_down_1": "cohort_group",
            "drill_down_2": "admission_source",
            "metric_month": "hospital_discharge_month",
            "num": "hospital_los_days",
            "denom": "null",
            "num_calculation": "'percentile_25'",
            "denom_calculation": "null",
            "metric_type": "'percentile'",
            "direction": "'down'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_visits')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Effective and Efficient Care'",
            "metric_name": "'75th percentile Hospital LOS'",
            "drill_down_1": "cohort_group",
            "drill_down_2": "admission_source",
            "metric_month": "hospital_discharge_month",
            "num": "hospital_los_days",
            "denom": "null",
            "num_calculation": "'percentile_75'",
            "denom_calculation": "null",
            "metric_type": "'percentile'",
            "direction": "'down'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_visits')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Effective and Efficient Care'",
            "metric_name": "'Harm Index'",
            "drill_down_1": "harm_type",
            "drill_down_2": "'No drilldown available'",
            "metric_month": "harm_event_dt_month",
            "num": "num_of_harm_events",
            "denom": "num_of_population_days",
            "num_calculation": "'sum'",
            "denom_calculation": "'sum'",
            "metric_type": "'rate'",
            "direction": "'down'",
            "month_target": "0",
            "annual_target": "0",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_harm_index')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Growth'",
            "metric_name": "'Number of ECMO runs'",
            "drill_down_1": "treatment_team_at_ecmo_start",
            "drill_down_2": "'No drilldown available'",
            "metric_month": "date_trunc('month', ecmo_end_datetime)",
            "num": "visit_key + cast(extract(epoch from ecmo_end_datetime) as int)",
            "denom": "null",
            "num_calculation": "'count'",
            "denom_calculation": "null",
            "metric_type": "'count'",
            "direction": "'up'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'count'",
            "metric_lag": "null",
            "table": source('stacks', 'neo_nicu_ecmo_run')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Growth'",
            "metric_name": "'Average length of ECMO run'",
            "drill_down_1": "cannulation_at_start_of_run",
            "drill_down_2": "'No drilldown available'",
            "metric_month": "date_trunc('month', ecmo_end_datetime)",
            "num": "ecmo_run_time_days",
            "denom": "null",
            "num_calculation": "'avg'",
            "denom_calculation": "null",
            "metric_type": "'avg'",
            "direction": "'down'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'rate'",
            "metric_lag": "null",
            "table": source('stacks', 'neo_nicu_ecmo_run')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Growth'",
            "metric_name": "'Number of episodes'",
            "drill_down_1": "cohort_group",
            "drill_down_2": "admission_source",
            "metric_month": "episode_end_month",
            "num": "episode_key_bigint",
            "denom": "null",
            "num_calculation": "'count'",
            "denom_calculation": "null",
            "metric_type": "'count'",
            "direction": "'up'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'count'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_episodes')
        },
        {
            "dashboard": "'neonatology service line dashboard'",
            "domain": "'Growth'",
            "metric_name": "'Number of surgical cases'",
            "drill_down_1": "service",
            "drill_down_2": "surgery_location_group",
            "metric_month": "surgery_month",
            "num": "or_key",
            "denom": "null",
            "num_calculation": "'count'",
            "denom_calculation": "null",
            "metric_type": "'count'",
            "direction": "'up'",
            "month_target": "null",
            "annual_target": "null",
            "metric_format": "'count'",
            "metric_lag": "null",
            "table": ref('stg_sl_dash_neo_surgical_cases')
        }
    ]
%}

{% set metrics = neo_metrics %}

with metric_data as (
    {%- for metric in metrics %}
    select
        {{ metric.dashboard }} as dashboard,
        {{ metric.domain }} as domain,
        {{ metric.metric_name }} as metric_name,
        {{ metric.drill_down_1 }} as drill_down_1,
        {{ metric.drill_down_2 }} as drill_down_2,
        {{ metric.metric_month }} as metric_month,
        cast({{ metric.num }} as numeric) as num,
        cast({{ metric.denom }} as numeric) as denom,
        {{ metric.num_calculation }} as num_calculation,
        {{ metric.denom_calculation }} as denom_calculation,
        {{ metric.metric_type }} as metric_type,
        {{ metric.direction }} as direction,
        cast({{ metric.month_target }} as numeric) as month_target,
        cast({{ metric.annual_target }} as numeric) as annual_target,
        {{ metric.metric_format }} as metric_format,
        {{ metric.metric_lag }} as metric_lag
    from
        {{ metric.table }}

    {% if not loop.last -%}
    union all
    {% endif %}

    {% endfor %}
)

select
    metric_data.*,
    calendar.*
from
    metric_data
    inner join {{ ref('stg_sl_dash_calendar') }} as calendar
        on metric_data.metric_month = calendar.visual_month
```

</ details>

alanmcruickshank commented:
Heads up team, that it might be a bit before I merge this. The excellent test query which @barrywhart produced has done a great job at flushing out quite a few bugs... which I now need to fix 🤦 . I don't know if I'll address all of the functionality in such a complex query, but I do want to make sure that we _fail safe_ at the very least [rather than what currently happens were it's unexpectedly removing large chunks of the file]. I'll report back when I've made progress on that front.

In the mean time - if people find any other oddities while testing - I'd love to hear them. Submitted hairy test queries are also still useful even if you've not had chance to test them yourself.

alanmcruickshank commented:
....I think I need to stop on this. It's just getting bigger.

Since the last main review I've added a few significant extras:
- A big rejig of the autofix testing to better handle tests with lots of rules. The names were getting really hard to reference and the new layout with config files is much simpler.
- Some additional fail safes including a total skip of handling templated placeholder indentation if we don't parse a balanced set of them (the same number of indents as dedents) in the source file. The reason it's non deterministic is because we're using some guesswork to understand what's going on within templated sections and occasionally that misses one. This seems particularly prevalent around nested for and if statements.
- We can now parse config lines which apply to the whole file directly from source comments. a la `-- sqlfluff: rules: L003: lint_templated_tokens: True`. More documentation is probably required around this and also some more tests, but I propose this happens in a seperate PR.
- Fixed a major bug with the position markers of fixes getting tangled up and some VERY STRANGE fix behaviour happening.
- As per the comments above, there's also a stub where I anticipate we could build out some limited fixes to the templated tokens (such as fixing `{{ref()}}` to `{{ ref() }}`). I haven't built this out, but I suggest we do that in a seperate PR.

I believe that even with some pretty hairy queries, this feature now at least _fails safe_ and as such I suggest we merge shortly. It's already quite a huge PR and I'm a bit worried that if I take it any further the merge conflicts are going to be significant.

Last call for any concerns on this? Otherwise I'll merge in the next few days.

pwildenhain commented:
Go for it 🤞

You've done a great job laying out what's left -- can we spin these out into new issues, referencing this PR (a la "Reference in new issue" functionality)

This was we don't lose track of what's left, and the context it was left in

alanmcruickshank commented:
> can we spin these out into new issues

Yes I'll do just that. I'd like to merge #508 before merging this (so that I can manage the considerable merge conflicts it will generate with this), so I'll make a note to make those new issues this week.
